Build separate rtld, libc, and libthr for sandboxing

brooksdavis opened the pull request:
Libraries are installed in /lib/c18n and /usr/lib/c18n. Debug symbols are also installed in the expected places (debug c18n directories are currently created unconditionally).

The runtime linker is installed as /libexec/ld-elf-c18n.so.1.

Currently a static libc.a is installed which probably isn't useful, but I failed to find an easy way to disable it.

I've not updated rtld to look for libraries in /lib/c18n yet.

dpgao committed:
Add forwarding-only trampoline

dpgao committed:
Add single-threaded returning trampoline

dpgao committed:
Add multi-threaded returning trampoline

dpgao committed:
Push rcsp onto executive mode stack in trampoline

dpgao committed:
Use trampoline at the end of rtld_start

dpgao committed:
Add stack-switching return trampoline

dpgao committed:
Add TLS support

dpgao committed:
Insert trampolines when calling dlsym

dpgao committed:
Fix trampoline bug where saved stack pointer is not restored to correct value

dpgao committed:
Executive stack frames now contain pointers linking to the next frame

dpgao committed:
Add support for setjmp/longjmp

dpgao committed:
Improve thread creation hook

dpgao committed:
Use non-hashtable-based stack lookup by default

dpgao committed:
Add weak stubs for _rtld_setjmp/_rtld_longjmp in libc

dpgao committed:
Add weak symbol to libthr

dpgao committed:
Overhaul hook functions between rtld and libc/libthr

jrtc27 commented on libexec/rtld-elf/Makefile:
```diff
@@ -30,7 +28,7 @@ MK_ASAN=	no
 MK_UBSAN=	no
 
-.if !defined(NEED_COMPAT)
+.if !defined(NEED_COMPAT) && !defined(RTLD_SANDBOX)
```
Not set by the Makefile

jrtc27 commented on libexec/rtld-elf-c18n/Makefile:
```diff
@@ -0,0 +1,12 @@
+MAN=
+MLINKS=	rtld.1 ld-elf-c18n.so.1.1
+
+CFLAGS+=-DRTLD_SANDBOX
```
Should be set by rtld-elf?

jrtc27 commented on libexec/rtld-elf-c18n/Makefile:
```diff
@@ -0,0 +1,12 @@
+
+CFLAGS+=-DRTLD_SANDBOX
+
+MK_TESTS:=	no
```
Maybe this too

arichardson commented on lib/libc/Makefile:
```diff
@@ -34,7 +36,11 @@ LIBC_ARCH=${MACHINE_CPUARCH}
 # To include RCS ID strings from other BSD projects, remove -DNO__RCSID from CFLAGS.
 CFLAGS+=-DNO__SCCSID -DNO__RCSID
 
+.ifdef RTLD_SANDBOCX
```
Typo?

dpgao committed:
Improve style

dpgao committed:
Remove hashtable-based trampoline

brooksdavis committed:
Install separate rtld sandbox implementations

Build rtld sandbox versions of libc and libthr and install in
/usr/lib/c18n. Libraries are not built for lib64c.

Also build and install an appropriately built runtime linker as
/libexec/ld-elf-c18n.so.1.

brooksdavis committed:
libc: fix rtld sandbox stub compilation

Only build for purecap and hoist the definitions in rtld_lock.h out of
the IN_RTLD block.

Declare arguments __unused.

brooksdavis commented:
I've updated things to fix issues pointed out by @jrtc27 and @arichardson as well as rebased. This fixed the creation of static versions of libc.

dpgao commented on lib/libc/gen/dlfcn.c:
```diff
@@ -192,17 +192,19 @@ _rtld_thread_init(void *li __unused)
 	/* Do nothing when linked statically. */
 }
 
+#ifdef __CHERI_PURE_CAPABILITY__
```
@brooksdavis I wonder if these stubs (together with the declarations in `rtld_lock.h`) should be wrapped in `#if defined(RTLD_SANDBOX)` so that non-c18n binaries do not have these extra symbols.

(Doing so would require `ifdef`s in `Symbol.map`, which I'm not sure is possible.)

brooksdavis replied:
I think I tried that and something failed to link, but my build tree may not have been sufficiently clean.

No ifdef is required in Symbol.map because it's not an error for a missing symbol to be listed (this is sometimes annoying). That being said, having a lib/libc/gen/Symbol-c18n.map or similar and adding it to SYM_MAPS conditionally in the Makefile could make sense.

dpgao commented:
> I've not updated rtld to look for libraries in /lib/c18n yet.

This can probably be done by adding the following to `rtld_paths.h`. (I have not tested this, though.)

```
#ifdef RTLD_SANDBOX
#define	STANDARD_LIBRARY_PATH	"/lib/casper:/lib/c18n:/lib:/usr/lib/c18n:/usr/lib"
#endif
```

jrtc27 commented:
Do we want `LD_` to be `LD_C18N_`?

brooksdavis commented:
> Do we want `LD_` to be `LD_C18N_`?

Yeah, I'll go ahead and make the full set of changes.

brooksdavis committed:
Allow MAP_FIXED mappings without CHERI_PERM_EXECUTIVE

If CHERI_PERM_EXECUTIVE has been stripped from the hint capability,
don't require it in the implied permissions. Without this, a sandboxed
rtld fails when mapping portions off a program PROT_READ|PROT_EXEC
because CHERI_PERM_EXECUTIVE has previous been stripped.  This also
follows if a wrapper stripps CHERI_PERM_EXECUTIVE from mmap() returns.

brooksdavis commented:
I've tried to test these changed in morello qemu. When I initially tried to direct exec a program I got mmap failures in rtld due to the stripping of CHERI_PERM_EXECUTIVE from the result returns from mmap before the followup MAP_FIXED calls. I've added a change to explicitly allow CHERI_PERM_EXECUTIVE to not be added in those cases.

Unfortunately, that isn't sufficient and I'm getting a crash due to under-aligned calls to __clear_cache() in tramp_pgs_append (note that while this is /bin/echo, I get the same crash with tmux):
```
Core was generated by `/libexec/ld-elf-c18n.so.1 /bin/echo foo'.
Program terminated with signal SIGPROT, CHERI protection violation
Capability bounds fault.
#0  __clear_cache (start=0x4015d034 [rwxRWE,0x4015d020-0x4015d0f0], 
    end=0x4015d0f0 [rwxRWE,0x4015d020-0x4015d0f0])
    at /home/bed22/cheri/cheribsd/contrib/subrepo-cheri-compiler-rt/lib/builtins/clear_cache.c:128
128     /home/bed22/cheri/cheribsd/contrib/subrepo-cheri-compiler-rt/lib/builtins/clear_cache.c: No such file or directory.
(gdb) bt
#0  __clear_cache (start=0x4015d034 [rwxRWE,0x4015d020-0x4015d0f0], 
    end=0x4015d0f0 [rwxRWE,0x4015d020-0x4015d0f0])
    at /home/bed22/cheri/cheribsd/contrib/subrepo-cheri-compiler-rt/lib/builtins/clear_cache.c:128
#1  0x0000000000118b60 in tramp_pgs_append (target=unsigned __intcap, 
    dst=0x40157810 [rwRW,0x40157810-0x40157d50])
    at /home/bed22/cheri/cheribsd/libexec/rtld-elf/aarch64/reloc.c:380
#2  0x000000000011ce4c in objlist_call_init (list=<optimized out>, 
    lockstate=<optimized out>)
    at /home/bed22/cheri/cheribsd/libexec/rtld-elf/rtld.c:3381
#3  0x000000000011b84c in _rtld (
    aux=0xffffbff7f9a0 [rwRW,0xffffbff7f9a0-0xffffbff7fe00], 
    exit_proc=0xfffffff7ffe0 [rwRW,0xfffffff7ffe0-0xfffffff7fff0], 
    objp=0xfffffff7fff0 [rwRW,0xfffffff7fff0-0xfffffff80000])
    at /home/bed22/cheri/cheribsd/libexec/rtld-elf/rtld.c:1092
#4  0x0000000000118448 in rtld_start ()
    at /home/bed22/cheri/cheribsd/libexec/rtld-elf/aarch64/rtld_start.S:70
```

dpgao commented:
> I've tried to test these changed in morello qemu.

`__clear_cache` does not seem to work on qemu. Try deleting that line and run again?

(This call works fine on Morello.)

jrtc27 commented:
It works fine on Morello because it has instruction-to-data and data-to-instruction coherence, i.e. CTR_EL0.DIC/IDC are set, meaning both the DC CVAU and IC IVAU instructions in __clear_cache get skipped as an optimisation.

On QEMU we don't report that coherence (it may well be safe to do though) and so those instructions are not skipped.

This is a bug in the Morello specification as it's overly-restrictive on the bounds of capabilities used for operations that should be safe.

jrtc27 commented:
In fact I believe the DC CVAU and IC IVAU are safe to skip; there is no coherence needed in QEMU other than to ensure the translation block is ended, but that corresponds more closely to pipeline barriers not cache maintenance instructions, i.e. the unconditional `isb sy` at the end (which definitely _is_ needed on both implementations, so you can't just remove the call).

dpgao commented on sys/vm/vm_mmap.c:
```diff
@@ -388,6 +388,14 @@ sys_mmap(struct thread *td, struct mmap_args *uap)
 
 	perms = cheri_getperm(source_cap);
 	reqperms = vm_map_prot2perms(uap->prot);
+#ifdef CHERI_PERM_EXECUTIVE
```
Reading the source code of `vm_map_prot2perms`, I cannot see how the return value can have `CHERI_PERM_EXECUTIVE` set. Am I missing something here?

brooksdavis replied:
Chase the set of perm ifdefs and you'll find:
```
#define CHERI_PERMS_PROT2PERM_EXEC                                      \
        (CHERI_PERM_EXECUTE | CHERI_PERM_EXECUTIVE |                    \
         CHERI_PERMS_PROT2PERM_READ)
```

brooksdavis commented:
Hmm, do we have a way to detect we're running on qemu or should we alter qemu to set CTR_EL0.DIC/IDC? It would be best if this work wasn't restricted to real hardware.

jrtc27 commented:
I mean, it's kind of useful that QEMU is telling you there's a bug here. The issue is you can't necessarily easily fix it... though in this case you could as you control the allocation and can align it. Ideally you'd therefore align and pad the trampoline allocation regions to a cache line so this code isn't relying on __clear_cache's optimisation. That or just replace the call to __clear_cache with a single explicit `isb sy` with a comment about why it's ok despite not being sufficient per the spec and why we're not using __clear_cache, but I don't see in this case why we can't just be a bit more careful about the allocations.

dpgao commented:
> align and pad the trampoline allocation regions to a cache line

@jrtc27 It seems that `__clear_cache` already alignes the address to the size of a cache line? See
```
  if (((ctr_el0 >> 28) & 0x1) == 0x0) {
    const size_t dcache_line_size = 4 << ((ctr_el0 >> 16) & 15);
    for (addr = xstart & ~(dcache_line_size - 1); addr < xend;
         addr += dcache_line_size)
      __asm __volatile("dc cvau, %0" ::"r"(addr));
  }
```
in `clear_cache.c`.

dpgao committed:
Align the capability bounds of the arguments of __clear_cache

dpgao commented:
It turns out that the alignment of the addresses passed to `__clear_cache` do not matter. ~But the alignment of the _bounds_ of these addresses do.~ But these addresses, after being aligned, must fall within the bounds passed capabilities. 8b74558043e046374290c2f58e6592d29c7232bf fixes the call to `__clear_cache` and now the linkers runs on QEMU fine.

jrtc27 commented:
Yes, that's why I said to align and pad the allocation, not just to align the pointer

dpgao committed:
Replace uintptr_t with void * where appropriate to reduce casts

dpgao committed:
Add cheribsdtest-purecap-dynamic-mt-c18n

dpgao commented:
I've updated `cheribsdtest` (e7100539102d889322f6aa9ea5e3028a480dee1e) with `cheribsdtest-purecap-dynamic-mt-c18n` that tests compartmentalisation-related features. So far I only removed/changed existing tests that fail incorrectly and haven't added new tests.

Still, the following two tests are failing on CheriBSD 22.05p1
```
  cheribsdtest_vm_reservation_mmap_fixed_unreserved: Call 'mmap((void *)hint, PAGE_SIZE, PROT_MAX(PROT_READ | PROT_WRITE), MAP_ANON | MAP_FIXED, -1, 0)' failed: Cannot allocate memory
  test_cheriabi_shmdt_invalid_ptr: Call 'shmat(shmid, requested_addr, 0)' failed: Cannot allocate memory
```

For `cheribsdtest_vm_reservation_mmap_fixed_unreserved`, it fails simply because the hint address is already in use.

For `test_cheriabi_shmdt_invalid_ptr`, it only fails when run with `-a`, and does not fail when run individually. I'm not sure how to diagnose this issue. But it is likely the same issue as above.

brooksdavis committed:
cheribsdtest: use a higher arbitrary hint

The library compartmentalization runtime creates more, larger mappings
and overlapped with the previous hint.  Shift the address up as a
workaround.  All of these tests need to be updated to use known free
memory, perhaps with an interface using libprocstat to find empty
address space.

dpgao commented:
Rebased onto CTSRD-CHERI/cheribsd#1498.
